### PR TITLE
Parse inline tuple types

### LINKED ISSUE #915: Parse inline tuple types
Support something like

```
    [int, int] p = [17, 42];
```

### PR DESCRIPTION
**Pull request by ushirask**
Fixes #915

### CODE REVIEW

File: compiler/modules/front.syntax/parseExpr.bal [hunk 1]
```diff
@@ -5,14 +5,7 @@ type SpecialMethodName "map";
 function parseExpr(Tokenizer tok) returns Expr|err:Syntax {
     Token? t = tok.current();
     Position startPos = tok.currentStartPos();
-    if t == "[" {
```
Comment (jclark) [inline compiler/modules/front.syntax/parseExpr.bal]: Why have you removed this?  Please carefully look at subset grammar and make sure you are parsing *exactly* that.  List constructors are allowed only at the top level.

File: compiler/modules/front.syntax/parseExpr.bal [hunk 2]
```diff
@@ -361,6 +354,13 @@ function startPrimaryExpr(Tokenizer tok) returns Expr|err:Syntax {
         endPos = check tok.expectEnd(")");
         return { startPos, endPos, message, kwPos };
     }
+    else if t == "[" {
```
Comment (jclark) [inline compiler/modules/front.syntax/parseExpr.bal]: No. This does not parse the grammar in the subset spec.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front.syntax/parseExpr.bal (hunk 1 above)
Comment (ushirask) [reply compiler/modules/front.syntax/parseExpr.bal]: I changed it earlier to parse `[X].method()` which is not allowed in the subset. Reverted and fixed it.

File: compiler/modules/front.syntax/preparse.bal [hunk 3]
```diff
@@ -16,7 +16,16 @@ final readonly & map<CLOSE_BRACKET> closeBracketMap = {
...
 function preparseParenTypeDesc(Tokenizer tok) returns boolean|err:Syntax {
-    boolean? parenResult = check preparseBracketed(tok, ")");
+    return preparseBracketTypeDesc(tok, ")");
+}
+
+function preparseTupleTypeDesc(Tokenizer tok) returns boolean|err:Syntax {
+    return preparseBracketTypeDesc(tok, "]");
+}
+
+function preparseBracketTypeDesc(Tokenizer tok, CLOSE_BRACKET close) returns boolean|err:Syntax {
```
Comment (jclark) [inline compiler/modules/front.syntax/preparse.bal]: preparseBracketedTypeDesc

File: compiler/modules/front.syntax/tests/testPreparse.bal [hunk 4]
```diff
@@ -22,7 +22,6 @@ function testPreparse(string src, PreparseTestResult expected, int minLookahead,
     TokenizerState state = tok.save();
 
     test:assertEquals(tok.current(), "(");
-    check tok.advance();
```
Comment (jclark) [inline compiler/modules/front.syntax/tests/testPreparse.bal:25]: Why is this changed needed?

**Review by jclark — COMMENTED** [review]
Also need way more tests..

File: compiler/modules/front.syntax/tests/testPreparse.bal (hunk 4 above)
Comment (ushirask) [reply compiler/modules/front.syntax/tests/testPreparse.bal:25]: Since `tok.advance()` was moved into `preparseParenTypeDesc` to accommodate for the `savePreparseRestore` helper function, we don't need to call it again here.

File: compiler/testSuite/14-tuple/inline1-v.bal [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import ballerina/io;
+
+type T 1|2|3;
+
+public function main() {
+    [int, int] p = [17, 42];
+    io:println(p); // @output [17,42]
+    [int, string, boolean] t1 = [2, "x", true];
+    io:println(t1); // @output [2,"x",true]
+    [T, string] t2 = [2, "test"];
```
Comment (jclark) [inline compiler/testSuite/14-tuple/inline1-v.bal:10]: More tests needed:

```
[T]
[int]
[T...]
[(int|float)]
[(T|float)]
```

etc.

**Review by jclark — CHANGES_REQUESTED** [review]
